### PR TITLE
[router] add base_gpu_id server args & merged radix tree python reference

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

This PR prepares for https://github.com/sgl-project/sglang/pull/2114 by:

1. Adding `base_gpu_id` server argument to set the device ID offset when starting TP server. This enables starting multiple TP servers with different base GPU IDs when launching several servers and the router together using a single command. This also eliminates the need of setting `CUDA_VISIBLE_DEVICE` if we want to launch on certain GPUs

2. Adding merged radix tree Python implementation. This serves as an experimental Python version which I referenced for my Rust implementation.

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [x] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: python/sglang/srt/server_args.py [hunk 1]
```diff
@@ -412,6 +413,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             default=ServerArgs.download_dir,
             help="Model download directory.",
         )
+        parser.add_argument(
+            "--base-gpu-id",
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py:417]: can we use CUDA_VISIBLE_DEVICES to achieve this?

Comment (ByronHsu) [reply python/sglang/srt/server_args.py:417]: https://github.com/sgl-project/sglang/pull/2114/files#diff-38af500b72b9f4e46bd0beafdbad7f7cfdbbe82a6dfd3284b18d7eb4948adb8eR35

Here is my current impl of the new `--dp-size`, which launches multiple TP server before the router starts. With base gpu id, the implementation is quite clean. Pretty much a decoupled version of the existing dp controller, where we also have a based gpu id.

 I have tried `CUDA_VISIBLE_DEVICES`, but it is not set correctly in the subprocess for some reasons. Another reason is that base-gpu-id can work with all torch supported device and `CUDA_VISIBLE_DEVICES` is only for NVIDIA.
 
 Any concern for adding this?

Comment (merrymercy) [reply python/sglang/srt/server_args.py:417]: ok, we can use this.

Can you also fix this for `--dp`?

Comment (ByronHsu) [reply python/sglang/srt/server_args.py:417]: Do you mean add `serving_args.base_gpu_id +` to https://github.com/sgl-project/sglang/blob/f6f713797bcbc63d225136d66deaa00495cdedfe/python/sglang/srt/managers/data_parallel_controller.py#L159?

**Comment by ByronHsu** [comment]
@merrymercy can you take another look?

File: python/sglang/srt/server_args.py (hunk 1 above)
Comment (zhyncs) [reply python/sglang/srt/server_args.py:417]: After setting `CUDA_VISIBLE_DEVICES`
e.g.
```
export CUDA_VISIBLE_DEVICES=4,6
```
Then the program can only see the 0 and 1. This IDs have been automatically rearranged. The 0 and 1 here actually refer to 4 and 6.

**Pull request merged**